pkg/operation/botanist/component/extensions/operatingsystemconfig/executor/templates/scripts/execute-cloud-config.tpl.sh

```diff
@@ -173,11 +173,11 @@ fi
 
 # Try to find Node object for this machine if already registered to the cluster.
 NODENAME=
-if [[ -s "{{ .pathNodeName }}" ]]; then
+if [[ -s "{{ .pathNodeName }}" ]] && [[ ! -z "$(cat "{{ .pathNodeName }}")" ]]; then
   NODENAME="$(cat "{{ .pathNodeName }}")"
 elif [[ -f "{{ .pathKubeletKubeconfigReal }}" ]]; then
   {{`NODENAME="$(`}}{{ .pathBinaries }}{{`/kubectl --kubeconfig="`}}{{ .pathKubeletKubeconfigReal }}{{`" get nodes -l "kubernetes.io/hostname=$(hostname | tr '[:upper:]' '[:lower:]')" -o go-template="{{ if .items }}{{ (index .items 0).metadata.name }}{{ end }}")"`}}
-  echo "$NODENAME" > "{{ .pathNodeName }}"
+  echo -n "$NODENAME" > "{{ .pathNodeName }}"
```

Comment:
But the `NODENAME` could be still empty, e.g. because of failure on the previous line.

Reply:
Yes, what is your concern? The next execution will try it again?

Reply:
There is skipped code when `NODENAME` is empty,
For example, the node will not get the new cloud-config checksum https://github.com/gardener/gardener/blob/10a2c08a5c89ca8c781fcb8f119afafe1723f6e0/pkg/operation/botanist/component/extensions/operatingsystemconfig/executor/templates/scripts/execute-cloud-config.tpl.sh#L260-L266

And generally, I think the script should exit on failures, but here it could be tricky as this could prevent bootstrapping of new nodes that will never get the `kubelet` service running because of such early exit. 

But looking again, this code should be executed on the next re-try when eventually the `NODENAME` will not be empty.